src/services/datasource.ts

```diff
@@ -25,26 +44,82 @@ class WrappedLokiDatasource extends RuntimeDataSource<DataQuery> {
       getDataSourceSrv()
         .get(getDataSource(request.scopedVars.__sceneObject.valueOf()))
         .then((ds) => {
-          // override the target datasource to Loki
-          request.targets = request.targets.map((target) => {
-            target.datasource = ds;
-            return target;
-          });
-
-          // query the datasource and return either observable or promise
-          const dsResponse = ds.query(request);
-          if (isObservable(dsResponse)) {
-            dsResponse.subscribe(subscriber);
-          } else {
-            dsResponse.then((response) => {
-              subscriber.next(response);
-              subscriber.complete();
-            });
+          if (!(ds instanceof DataSourceWithBackend)) {
+            throw new Error('Invalid datasource!');
+          }
+
+          const requestType = request.targets?.[0]?.resource;
+
+          switch (requestType) {
+            case 'volume': {
+              this.transformVolumeResponse(request, ds, subscriber);
+              break;
+            }
+            case 'patterns': {
+              console.warn('not yet implemented');
+              // this.transformPatternResponse(request, ds, subscriber);
+              break;
+            }
+            default: {
+              // override the target datasource to Loki
+              request.targets = request.targets.map((target) => {
+                target.datasource = ds;
+                return target;
+              });
+
+              // query the datasource and return either observable or promise
+              const dsResponse = ds.query(request);
+              dsResponse.subscribe(subscriber);
+              break;
+            }
           }
         });
     });
   }
 
+  private transformVolumeResponse(
```

Comment:
Maybe rename this function into something like `requestVolume` / `getVolume`? I think `transformResponse` is just a "minor" part of this method.